exchanges/orderbook/Tranches.go

```diff
@@ -0,0 +1,669 @@
+package orderbook
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/thrasher-corp/gocryptotrader/common/math"
+)
+
+// FullLiquidityExhaustedPercentage defines when a book has been completely
+// wiped out of potential liquidity.
+const FullLiquidityExhaustedPercentage = -100
+
+var (
+	errIDCannotBeMatched               = errors.New("cannot match ID")
+	errCollisionDetected               = errors.New("cannot insert update, collision detected")
+	errAmountCannotBeLessOrEqualToZero = errors.New("amount cannot be less than or equal to zero")
+	errInvalidNominalSlippage          = errors.New("invalid slippage amount, its value must be greater than or equal to zero")
+	errInvalidImpactSlippage           = errors.New("invalid slippage amount, its value must be greater than zero")
+	errInvalidSlippageCannotExceed100  = errors.New("invalid slippage amount, its value cannot exceed 100%")
+	errBaseAmountInvalid               = errors.New("invalid base amount")
+	errInvalidReferencePrice           = errors.New("invalid reference price")
+	errQuoteAmountInvalid              = errors.New("quote amount invalid")
+	errInvalidCost                     = errors.New("invalid cost amount")
+	errInvalidAmount                   = errors.New("invalid amount")
+	errInvalidHeadPrice                = errors.New("invalid head price")
+	errNoLiquidity                     = errors.New("no liquidity")
+)
+
+// Tranches defines a slice of orderbook Tranche
+type Tranches []Tranche
+
+// comparison defines expected functionality to compare between two reference
+// price levels
+type comparison func(float64, float64) bool
+
+// load iterates across new tranches and refreshes stored slice with this
+// incoming snapshot.
+func (ts *Tranches) load(incoming Tranches) {
+	if len(incoming) == 0 {
+		*ts = (*ts)[:0] // Flush
+		return
+	}
+	if len(incoming) <= len(*ts) {
+		copy(*ts, incoming)         // Reuse
+		*ts = (*ts)[:len(incoming)] // Flush excess
+		return
+	}
+	if len(incoming) > cap(*ts) {
+		*ts = make([]Tranche, len(incoming)) // Extend
+		copy(*ts, incoming)                  // Copy
+		return
+	}
+	*ts = (*ts)[:0]                // Flush
+	*ts = append(*ts, incoming...) // Append
```

Comment:
I'm finding your cap copy faster than the Flush append
```go
func BenchmarkCopying1(b *testing.B) {
	b.ReportAllocs()
	for i := 0; i < b.N; i++ {
		one := []Tranche{{}, {}, {}, {}, {}, {}}
		two := []Tranche{{}, {}, {}}
		two = two[:0]
		two = append(two, one...)
	}
}

func BenchmarkCopying2(b *testing.B) {
	b.ReportAllocs()
	for i := 0; i < b.N; i++ {
		one := []Tranche{{}, {}, {}, {}, {}, {}}
		two := []Tranche{{}, {}, {}}
		two = make([]Tranche, len(one))
		copy(two, one)
	}
}
```

```
BenchmarkCopying1-10    	 7716270	       152.4 ns/op	     480 B/op	       1 allocs/op
BenchmarkCopying2-10    	 8571086	       144.6 ns/op	     480 B/op	       1 allocs/op

Reply:
indeed, what a guy!

Reply:
The only downside I think to that, is when two is bigger than one, and you copy over then you lose a bunch of data. @gloriousCode Confirm?